### PR TITLE
bug-fix for infinite loop in MuonSimClassifier.cc

### PR DESCRIPTION
**Pull request by Fedespring**
#### PR description:

Infinite loop happens in the RECOSIM step, because Gen Muons appear as the first GenParticle in the tree: https://its.cern.ch/jira/browse/CMSMUONS-312
Added a break to end the loop (when it happens) still keeping the LogTrace.
This fix the issue.

#### PR validation:
Validation done using the configuration, LHE and few root files provided by pdmv (can be found in
/afs/cern.ch/user/s/srimanob/public/ForPdmV/McM/PPS-RunIIFall18GS-00004)

Running runTheMatrix.py -l limited -i all
6 5 4 3 2 0 0 0 0 tests passed, 28 0 0 0 0 0 0 0 0 failed 

Backporting is needed in 10_2_X

### CODE REVIEW

**Comment by srimanob** [comment]
@Fedespring 
Thanks. Local test seems to run fine in MiniAOD step. Another following issue in NanoAOD is reported in https://github.com/cms-sw/cmssw/issues/28803
Somehow rivert can match only a single parton.

**Comment by civanch** [comment]
please test

File: SimMuon/MCTruth/plugins/MuonSimClassifier.cc [hunk 1]
```diff
@@ -320,9 +320,11 @@ void MuonSimClassifier::produce(edm::Event &iEvent, const edm::EventSetup &iSetu
               jm++;
               if (mMom->numberOfMothers() > 0) {
                 mMom = mMom->motherRef();
+              } else {
+                LogTrace("MuonSimClassifier") << "\t\t backtracking mother " << jm << ", pdgId = " << mMom->pdgId()
+                                              << ", status= " << mMom->status();
```
Comment (silviodonato) [inline SimMuon/MCTruth/plugins/MuonSimClassifier.cc]: Why do you move the LogTrace inside the "else"?

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
+1

**Comment by silviodonato** [comment]
+1

**Comment by srimanob** [comment]
Hi @Fedespring 

Do I understand correctly that backport to 10_2 has not happed yet?

I make the backport at
https://github.com/cms-sw/cmssw/pull/29065

It is needed for MC production.

Thx.

**Comment by Fedespring** [comment]
Yes correct.
Thanks

Il Sab 29 Feb 2020, 22:41 Phat Srimanobhas <notifications@github.com> ha
scritto:

**Pull request merged**